internal/grpc/services/publicstorageprovider/publicstorageprovider.go

```diff
@@ -458,48 +462,95 @@ func (s *service) Stat(ctx context.Context, req *provider.StatRequest) (*provide
+		parts := strings.Split(req.Ref.ResourceId.OpaqueId, "/")
+		tkn = parts[0]
+		nodeID = parts[1]
+	} else if req.Ref.Path != "" {
+		var err error
+		tkn, relativePath, err = s.unwrap(ctx, req.Ref)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	originalPath, ls, ri, st, err := s.resolveToken(ctx, tkn)
+	share, shareInfo, st, err := s.resolveToken(ctx, tkn)
 	switch {
 	case err != nil:
 		return nil, err
 	case st != nil:
 		return &provider.StatResponse{
 			Status: st,
 		}, nil
-	case ls.GetPermissions() == nil || !ls.GetPermissions().Permissions.Stat:
+	case share.GetPermissions() == nil || !share.GetPermissions().Permissions.Stat:
 		return &provider.StatResponse{
 			Status: status.NewPermissionDenied(ctx, nil, "share does not grant Stat permission"),
 		}, nil
 	}
 
-	p := originalPath
-	if ri.Type == provider.ResourceType_RESOURCE_TYPE_CONTAINER {
-		p = path.Join("/", p, relativePath)
+	if shareInfo.Type == provider.ResourceType_RESOURCE_TYPE_FILE {
+		res := &provider.StatResponse{
+			Status: status.NewOK(ctx),
+			Info:   shareInfo,
+		}
+		s.augmentStatResponse(ctx, res, shareInfo, share, tkn)
+		return res, nil
 	}
-	var statResponse *provider.StatResponse
-	// the call has to be made to the gateway instead of the storage.
-	statResponse, err = s.gateway.Stat(ctx, &provider.StatRequest{Ref: &provider.Reference{Path: p}})
+
+	var ref *provider.Reference
+	if req.Ref.ResourceId != nil {
+		ref = &provider.Reference{ResourceId: &provider.ResourceId{
+			StorageId: share.ResourceId.StorageId,
+			OpaqueId:  nodeID,
+		}}
+	} else if req.Ref.Path != "" {
+		p := path.Join("/", shareInfo.Path, relativePath)
+		ref = &provider.Reference{Path: p}
+	}
+
+	statResponse, err := s.gateway.Stat(ctx, &provider.StatRequest{Ref: ref})
 	if err != nil {
 		return &provider.StatResponse{
 			Status: status.NewInternal(ctx, err, "gateway: error calling Stat for ref:"+req.Ref.String()),
 		}, nil
 	}
 
+	s.augmentStatResponse(ctx, statResponse, shareInfo, share, tkn)
+
+	return statResponse, nil
+}
+
+func (s *service) augmentStatResponse(ctx context.Context, res *provider.StatResponse, shareInfo *provider.ResourceInfo, share *link.PublicShare, tkn string) {
 	// prevent leaking internal paths
-	if statResponse.Info != nil {
-		if err := addShare(statResponse.Info, ls); err != nil {
-			appctx.GetLogger(ctx).Error().Err(err).Interface("share", ls).Interface("info", statResponse.Info).Msg("error when adding share")
+	if res.Info != nil {
+		if err := addShare(res.Info, share); err != nil {
+			appctx.GetLogger(ctx).Error().Err(err).Interface("share", share).Interface("info", res.Info).Msg("error when adding share")
 		}
-		statResponse.Info.Path = path.Join(s.mountPath, "/", tkn, relativePath)
-		filterPermissions(statResponse.Info.PermissionSet, ls.GetPermissions().Permissions)
+
+		var sharePath string
+		if shareInfo.Type == provider.ResourceType_RESOURCE_TYPE_FILE {
+			sharePath = path.Base(shareInfo.Path)
```

Comment:
One question, but I might be misunderstanding this. `sharePath` should be empty for file shares right? Because `/$mountID/$token` will resolve to the file itself. If you then append the file name to the path obtained from the token, it won't exist.

Reply:
Yes, you're right. Now I'm wondering why it worked so far. :thinking:

Reply:
Appending the filename to the path is necessary otherwise something breaks. If you don't want to have the filename in there then I need to dig into why it's required. But without it other things like downloading shared files fail.

Reply:
No it's fine by me. I don't understand why it worked previously. Now because of the check on line 499, we return early for file shares. But if we didn't you'd essentially be statting `path-to-parent-dir/file.txt/file.txt` in line 519. Merging for now but please look into it if you have the time. The issue might be on the `web` end.